File: C++/Problems/Problems/problem_61.cpp
#include "problem_61.h"

bool problem_61::test() {
    auto node5 = ListNode(5);
    auto node4 = ListNode(4, &node5);
    auto node3 = ListNode(3, &node4);
    auto node2 = ListNode(2, &node3);
    auto head =  ListNode(1,  &node2);

    auto k = 2;

    auto node5e = ListNode(3);
    auto node4e = ListNode(2, &node5e);
    auto node3e = ListNode(1, &node4e);
    auto node2e = ListNode(5, &node3e);
    ListNode expected =  ListNode(4,  &node2e);

    auto result = rotateRight(&head, k);

    auto it = &expected;
    while (it != nullptr && result != nullptr)
    {
        if (it->val != result->val)
            return false;
        it = it->next;
        result = result->next;
    }

    return it == nullptr && result == nullptr;
}

problem_61::ListNode * problem_61::rotateRight(ListNode *head, int k) {
    if (head == nullptr)
        return nullptr;
    if (k == 0)
        return head;
    auto tail = head;
    int n = 1;
    while (tail->next != nullptr)
    {
        tail = tail->next;
        n++;
    }
    k = k % n;
    k = n - k - 1;
    auto b = head;
    while (k > 0)
    {
        b = b->next;
        k--;
    }
    tail->next = head;
    auto nHead = b->next;
    b->next = nullptr;

    return nHead;
}

File: Python/Problems/Problems/problem61.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Problem61(Problem):
    def test(self) -> bool:
        head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
        k = 2

        expected = ListNode(4, ListNode(5, ListNode(1, ListNode(2, ListNode(3)))))

        result = self.rotateRight(head, k)

        while expected != None and result != None:
            if expected.val != result.val:
                return False
            expected = expected.next
            result = result.next

        return expected == None and result == None

    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head == None:
            return None
        if k == 0:
            return head
        tail = head
        n = 1
        while (tail.next != None):
            tail = tail.next
            n += 1
        k = k % n
        k = n - k - 1
        b = head
        while k > 0:
            b = b.next
            k -= 1
        tail.next = head
        nHead = b.next
        b.next = None

        return nHead